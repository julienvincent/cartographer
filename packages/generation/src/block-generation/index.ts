export * from './block-space';
export * from './staircase';
