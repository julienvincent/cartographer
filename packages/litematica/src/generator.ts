import * as litematica_bit_array from './litematica-bit-array';
import * as _ from 'lodash';

type PaletteBlockProperty = {
  type: 'string';
  value: string;
};
type PaletteBlockProperties = {
  type: 'compound';
  value: Record<string, PaletteBlockProperty>;
};
type PaletteBlock = {
  Name: {
    type: 'string';
    value: string;
  };
  Properties?: PaletteBlockProperties;
};

export type Block = {
  id: string;
  properties?: Record<string, string>;
};

export type Coords = {
  x: number;
  y: number;
  z: number;
};

export type BlockPosition = Block & Coords;

const createPaletteBlock = (block: Block): PaletteBlock => {
  const palette_block: PaletteBlock = {
    Name: {
      type: 'string',
      value: block.id
    }
  };

  if (Object.keys(block.properties || {}).length > 0) {
    palette_block.Properties = {
      type: 'compound',
      value: _.mapValues(block.properties, (value) => {
        return {
          type: 'string',
          value: value
        } as PaletteBlockProperty;
      })
    };
  }

  return palette_block;
};

const comparePaletteBlocks = (a: PaletteBlock, b: PaletteBlock) => {
  if (a.Name.value !== b.Name.value) {
    return false;
  }

  if (a.Properties) {
    const keys = Object.keys(a.Properties.value);
    if (!b.Properties || Object.keys(b.Properties.value).length !== keys.length) {
      return false;
    }
    return keys.reduce((matching, key) => {
      if (!matching) {
        return false;
      }
      return a.Properties!.value[key].value === b.Properties!.value[key]?.value;
    }, true);
  } else {
    if (b.Properties) {
      return false;
    }
  }

  return true;
};

const createIndexKeyFromCoords = (block: Coords) => {
  return `${block.x}:${block.y}:${block.z}`;
};

const max = <T>(coll: T[], accessor: (item: T) => number) => {
  return coll.reduce((acc, item) => {
    const val = accessor(item);
    if (val > acc) {
      return val;
    }
    return acc;
  }, accessor(coll[0]));
};

type Params = {
  blocks: BlockPosition[];
  name?: string;
  description?: string;
  author?: string;
};
export const generateSchematicNBT = (params: Params) => {
  const length = max(params.blocks, (position) => position.z) + 1;
  const width = max(params.blocks, (position) => position.x) + 1;
  const height = max(params.blocks, (position) => position.y) + 1;

  const volume = length * width * height;
  const AIR = createPaletteBlock({
    id: 'minecraft:air'
  });

  const palette = params.blocks.reduce(
    (palette, block) => {
      const palette_block = createPaletteBlock(block);
      const exists = palette.find((existing) => {
        return comparePaletteBlocks(palette_block, existing);
      });
      if (!exists) {
        palette.push(palette_block);
      }
      return palette;
    },
    [AIR]
  );

  const index = new Map<string, Block>(
    params.blocks.map((block) => {
      return [createIndexKeyFromCoords(block), block];
    })
  );

  const bit_array = _.range(width).reduce((bit_array, x) => {
    return _.range(length).reduce((bit_array, z) => {
      return _.range(height).reduce((bit_array, y) => {
        const block = index.get(createIndexKeyFromCoords({ x, y, z }));

        const palette_index = palette.findIndex((item) => {
          if (block) {
            return comparePaletteBlocks(item, createPaletteBlock(block));
          }
          return comparePaletteBlocks(item, AIR);
        });

        const block_coords = (y * length + z) * width + x;
        return litematica_bit_array.set(bit_array, block_coords, palette_index);
      }, bit_array);
    }, bit_array);
  }, litematica_bit_array.createBitArray(volume, palette.length));

  return {
    name: params.name ?? '',
    value: {
      MinecraftDataVersion: {
        type: 'int',
        value: 2584
      },
      Version: {
        type: 'int',
        value: 5
      },
      Metadata: {
        type: 'compound',
        value: {
          TimeCreated: {
            type: 'long',
            value: [375, 2023476248] // TODO: Use correct time
          },
          TimeModified: {
            type: 'long',
            value: [375, 2023476248] // TODO: Use correct time
          },
          EnclosingSize: {
            type: 'compound',
            value: {
              x: {
                type: 'int',
                value: width
              },
              y: {
                type: 'int',
                value: height
              },
              z: {
                type: 'int',
                value: length
              }
            }
          },
          Description: {
            type: 'string',
            value: params.description ?? ''
          },
          RegionCount: {
            type: 'int',
            value: 1
          },
          TotalBlocks: {
            type: 'int',
            value: palette.length - 1
          },
          Author: {
            type: 'string',
            value: params.author ?? ''
          },
          TotalVolume: {
            type: 'int',
            value: volume
          },
          Name: {
            type: 'string',
            value: params.name ?? ''
          }
        }
      },
      Regions: {
        type: 'compound',
        value: {
          map: {
            type: 'compound',
            value: {
              BlockStates: {
                type: 'longArray',
                value: bit_array.array
              },
              PendingBlockTicks: {
                type: 'list',
                value: {
                  type: 'end',
                  value: []
                }
              },
              Position: {
                type: 'compound',
                value: {
                  x: {
                    type: 'int',
                    value: 1
                  },
                  y: {
                    type: 'int',
                    value: 0
                  },
                  z: {
                    type: 'int',
                    value: 1
                  }
                }
              },
              BlockStatePalette: {
                type: 'list',
                value: {
                  type: 'compound',
                  value: palette
                }
              },
              Size: {
                type: 'compound',
                value: {
                  x: {
                    type: 'int',
                    value: width
                  },
                  y: {
                    type: 'int',
                    value: height
                  },
                  z: {
                    type: 'int',
                    value: length
                  }
                }
              },
              PendingFluidTicks: {
                type: 'list',
                value: {
                  type: 'end',
                  value: []
                }
              },
              TileEntities: {
                type: 'list',
                value: {
                  type: 'end',
                  value: []
                }
              },
              Entities: {
                type: 'list',
                value: {
                  type: 'end',
                  value: []
                }
              }
            }
          }
        }
      }
    }
  };
};
