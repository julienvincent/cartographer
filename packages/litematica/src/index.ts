export * as litematica_bit_array from './litematica-bit-array';
export * as long from './long';
export * from './generator';
