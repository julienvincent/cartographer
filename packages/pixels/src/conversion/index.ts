export * from './pixel-data';
export * from './mc-blocks';
export * from './colors';
