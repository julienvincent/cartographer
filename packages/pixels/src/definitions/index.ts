export * from './palette';
export * from './block';
export * from './pixel';
