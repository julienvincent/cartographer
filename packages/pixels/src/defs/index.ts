export type RGBColor = [number, number, number];

export type Pixel = {
  r: number;
  g: number;
  b: number;
};

export type PixelGrid = Pixel[][];

export type BlockAttributes = Record<string, string> & {
  requires_support?: boolean;
  flammable?: boolean;
  liquid?: boolean;
};

export type MCBlockDefinition = {
  /**
   * The minecraft block ID. This will be something like `minecraft:air`
   */
  id: string;

  /**
   * Stores useful attributes about the block that can be used during the schema block space generation
   */
  attributes?: BlockAttributes;

  /**
   * Stores properties about how the block should be placed. This could be things like `{axis: 'y'}` on a
   * directional block such as `minecraft:birch_log` to indicate the direction the block is facing.
   */
  properties?: Record<string, string>;
};

export type MCBlockWithOffset = MCBlockDefinition & {
  y_offset: number;
};

export type MCBlockGrid = MCBlockWithOffset[][];

export type BlockPaletteItem = MCBlockDefinition & {
  colors: RGBColor[];
};
export type BlockPalette = BlockPaletteItem[];

export type ColorPaletteItem = {
  id: string;
  colors: RGBColor[];
  blocks: MCBlockDefinition[];
};
export type ColorPalette = ColorPaletteItem[];

export enum BlockColorSpectrum {
  Full = 'full',
  Flat = 'flat'
}

export type BlockPaletteItem = MCBlock & {
  colors: PaletteColorTuple;
};
export type BlockPalette = BlockPaletteItem[];
