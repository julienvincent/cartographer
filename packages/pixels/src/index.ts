export * as transformers from './transformers';
export * as conversion from './conversion';
export * as data from './data';
export * as defs from './defs';
