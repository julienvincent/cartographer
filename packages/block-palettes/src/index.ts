import { BlockPalette } from '@cartographer/pixels';

import palette_1_20 from './palettes/1.20.json';
import palette_1_19 from './palettes/1.19.json';
import palette_1_18 from './palettes/1.18.json';

export const palettes = {
  '1.20': palette_1_20 as unknown as BlockPalette,
  '1.19': palette_1_19 as unknown as BlockPalette,
  '1.18': palette_1_18 as unknown as BlockPalette
};
