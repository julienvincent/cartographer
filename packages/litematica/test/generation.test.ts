import { describe, it, expect } from 'vitest';

import * as litematica from '../src';
import * as _ from 'lodash';

describe('schematic-generation', () => {
  it('should generate a schematic file', () => {
    const data = litematica.generateSchematicNBT({
      blocks: [
        {
          id: 'minecraft:stone',
          x: 0,
          y: 0,
          z: 0
        },
        {
          id: 'minecraft:birch_log',
          properties: {
            axis: 'y'
          },
          x: 0,
          y: 1,
          z: 0
        },
        {
          id: 'minecraft:white_wool',
          x: 0,
          y: 0,
          z: 1
        },
        {
          id: 'minecraft:birch_log',
          properties: {
            axis: 'y'
          },
          x: 1,
          y: 0,
          z: 1
        },
        {
          id: 'minecraft:birch_log',
          properties: {
            axis: 'x'
          },
          x: 1,
          y: 1,
          z: 1
        }
      ]
    });

    expect(data).toMatchSnapshot();
  });

  it('should handle a large amount of data', () => {
    const num = 128 * 10;

    const blocks = _.range(num)
      .map((x) => {
        return _.range(num).map((z) => {
          return {
            id: 'minecraft:stone',
            x,
            y: 0,
            z
          };
        });
      })
      .flat();

    litematica.generateSchematicNBT({
      blocks: blocks
    });
  });
});
