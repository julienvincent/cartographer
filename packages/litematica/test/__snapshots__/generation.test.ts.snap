// Vitest Snapshot v1

exports[`schematic-generation > should generate a schematic file 1`] = `
{
  "name": "",
  "value": {
    "Metadata": {
      "type": "compound",
      "value": {
        "Author": {
          "type": "string",
          "value": "",
        },
        "Description": {
          "type": "string",
          "value": "",
        },
        "EnclosingSize": {
          "type": "compound",
          "value": {
            "x": {
              "type": "int",
              "value": 2,
            },
            "y": {
              "type": "int",
              "value": 2,
            },
            "z": {
              "type": "int",
              "value": 2,
            },
          },
        },
        "Name": {
          "type": "string",
          "value": "",
        },
        "RegionCount": {
          "type": "int",
          "value": 1,
        },
        "TimeCreated": {
          "type": "long",
          "value": [
            375,
            2023476248,
          ],
        },
        "TimeModified": {
          "type": "long",
          "value": [
            375,
            2023476248,
          ],
        },
        "TotalBlocks": {
          "type": "int",
          "value": 4,
        },
        "TotalVolume": {
          "type": "int",
          "value": 8,
        },
      },
    },
    "MinecraftDataVersion": {
      "type": "int",
      "value": 2584,
    },
    "Regions": {
      "type": "compound",
      "value": {
        "map": {
          "type": "compound",
          "value": {
            "BlockStatePalette": {
              "type": "list",
              "value": {
                "type": "compound",
                "value": [
                  {
                    "Name": {
                      "type": "string",
                      "value": "minecraft:air",
                    },
                  },
                  {
                    "Name": {
                      "type": "string",
                      "value": "minecraft:stone",
                    },
                  },
                  {
                    "Name": {
                      "type": "string",
                      "value": "minecraft:birch_log",
                    },
                    "Properties": {
                      "type": "compound",
                      "value": {
                        "axis": {
                          "type": "string",
                          "value": "y",
                        },
                      },
                    },
                  },
                  {
                    "Name": {
                      "type": "string",
                      "value": "minecraft:white_wool",
                    },
                  },
                  {
                    "Name": {
                      "type": "string",
                      "value": "minecraft:birch_log",
                    },
                    "Properties": {
                      "type": "compound",
                      "value": {
                        "axis": {
                          "type": "string",
                          "value": "x",
                        },
                      },
                    },
                  },
                ],
              },
            },
            "BlockStates": {
              "type": "longArray",
              "value": [
                [
                  1,
                  8398017,
                ],
              ],
            },
            "Entities": {
              "type": "list",
              "value": {
                "type": "end",
                "value": [],
              },
            },
            "PendingBlockTicks": {
              "type": "list",
              "value": {
                "type": "end",
                "value": [],
              },
            },
            "PendingFluidTicks": {
              "type": "list",
              "value": {
                "type": "end",
                "value": [],
              },
            },
            "Position": {
              "type": "compound",
              "value": {
                "x": {
                  "type": "int",
                  "value": 1,
                },
                "y": {
                  "type": "int",
                  "value": 0,
                },
                "z": {
                  "type": "int",
                  "value": 1,
                },
              },
            },
            "Size": {
              "type": "compound",
              "value": {
                "x": {
                  "type": "int",
                  "value": 2,
                },
                "y": {
                  "type": "int",
                  "value": 2,
                },
                "z": {
                  "type": "int",
                  "value": 2,
                },
              },
            },
            "TileEntities": {
              "type": "list",
              "value": {
                "type": "end",
                "value": [],
              },
            },
          },
        },
      },
    },
    "Version": {
      "type": "int",
      "value": 5,
    },
  },
}
`;
